File: app-examples/example-07-kmp/androidApp/src/main/kotlin/com/kmpfoo/android/OG.kt
package com.kmpfoo.android

import android.app.Application
import android.content.pm.ApplicationInfo
import co.early.fore.core.delegate.DebugDelegateDefault
import co.early.fore.core.delegate.Fore
import co.early.persista.PerSista
import com.kmpfoo.feature.counter.CounterModel
import okio.Path.Companion.toOkioPath

/**
 * Copyright © 2015-2023 early.co. All rights reserved.
 */
@Suppress("UNUSED_PARAMETER")
object OG {

    private var initialized = false
    private val dependencies = HashMap<Class<*>, Any>()

    fun setApplication(application: Application) {

        val isDebug = application.applicationInfo.flags and ApplicationInfo.FLAG_DEBUGGABLE != 0

        // create dependency graph
        if (isDebug) {
            Fore.setDelegate(DebugDelegateDefault("foo_"))
        }
        val logger = Fore.getLogger()

        val perSista = PerSista(
            dataPath = application.filesDir.toOkioPath(),
            logger = logger,
        )

        val counterModel = CounterModel(
            perSista = perSista,
            logger = logger,
        )

        // add models to the dependencies map if you will need them later
        dependencies[CounterModel::class.java] = counterModel
    }

    fun init() {
        if (!initialized) {
            initialized = true

            // run any necessary initialization code once object graph has been created here

//            // just in case you need to, but defaults are recommended
//            ForeSize.overrideBreakPoints(
//                ViewPortBreakPoints(
//                    widthSDpBelow = 400.dp,
//                    widthMDpBelow = 500.dp,
//                )
//            )

        }
    }

    /**
     * This is how dependencies get injected, typically an Activity/Fragment/View will call this
     * during the onCreate()/onCreateView()/onFinishInflate() method respectively for each of the
     * dependencies it needs.
     *
     * Can use a DI library for similar behaviour using annotations
     *
     * Will return mocks if they have been set previously in putMock()
     *
     *
     * Call it like this:
     *
     * <code>
     *     yourModel = OG[YourModel::class.java]
     * </code>
     *
     * If you want to more tightly scoped object, one way is to pass a factory class here and create
     * an instance where you need it
     *
     */
    @Suppress("UNCHECKED_CAST")
    operator fun <T> get(model: Class<T>): T = dependencies[model] as T

    fun <T> putMock(clazz: Class<T>, instance: T) {
        dependencies[clazz] = instance as Any
    }
}

File: app-examples/example-07-kmp/iosApp/iosApp/OG.swift
import Foundation
import shared

/*
 get dependencies from anywhere like this counterModel = OG[CounterModel.self]
 */
final class OG {

    private static let instance = OG()
    private var dependencies: [String: Any] = [:]
    
    private var initialized = false

    private init() {}

    static func create() {
        
        let logger = MultiplatformLogger()
        let fileManager = FileManager.default
        let documentsDirectory = fileManager.urls(for: .documentDirectory, in: .userDomainMask).first!
        let path = Path.Companion.shared.toPath(documentsDirectory.path, normalize: true)
        let perSista = PerSista(
            dataPath: path,
            logger: logger
        )
        
        // observable models
        let counterModel = CounterModel(perSista: perSista, logger: logger)
        
        // register dependencies for later
        instance.register(counterModel, as: CounterModel.self)
    }
    
    static func initialize() {
        if (!instance.initialized) {
            instance.initialized = true

            // run any necessary initialization code once object graph has been created here

        }
    }

    private func register<T>(_ dependency: T, as type: T.Type) {
        dependencies[String(describing: type)] = dependency
    }
    
    static subscript<T>(type: T.Type) -> T {
        guard let dependency = instance.dependencies[String(describing: type)] as? T else {
            fatalError("No dependency registered for type \(type)")
        }
        return dependency
    }

    func putMock<T>(_ instance: T, as type: T.Type) {
        OG.instance.register(instance, as: type)
    }
}
